### PR TITLE
CB-17317: Enable Datalake DR for the datalake service permanently, enabling integration tests to perform DR actions. Specifically, resize e2e test is extended to test for expected DR actions.

### PR DESCRIPTION
**Pull request by sxxgrc**
Jira: https://jira.cloudera.com/browse/CB-17317

The primary objective of this is to enable DR testing during resize integration tests, allowing us to ensure future changes do not break resize as a whole. Previously this was skipped during the tests, which was undesirable as it is not a true full e2e test.

Note that by enabling DR for all of the Datalake service, the only expected consequence is that local runs of flows which call the DR flow will not skip the process.

### CODE REVIEW

**Comment by sxxgrc** [comment]
New change: Removed the ability for the mock cloud provider to run backup/restore during resize as this is not currently allowed on Thunderhead.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxBackupRestoreService.java
Comment (horadla23) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxBackupRestoreService.java:535]: is there a jira for TODO?

Comment (horadla23) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxBackupRestoreService.java:535]: also I do not get this change completely. in the title it is said that you are enabling datalakedr for integration tests and e2e tests, though you are specifying here that in case of mock (which is the platform for integrations tests) backup should not be performed

**Review by RokolyA94 — APPROVED** [review]

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxBackupRestoreService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxBackupRestoreService.java:535]: Yes well the exact goal of this was to enable backups for the non-mock e2e tests using cloud providers such as AWS or Azure. 

However I have learned some new information which may allow me to remove this change here and add some other changes which gets the mock test working as well.

**Comment by sxxgrc** [comment]
Heavily modified this change to include the mock resize tests as well. Specifically I moved the validation to a central class used by both e2e tests, and included modifications to have the mock tests use the mock datalake DR service.

Despite this all passing for me locally I expect this to fail the integration tests here due to the build not running the DR service. Hopefully this can be fixed soon.

**Comment by sxxgrc** [comment]
I am blocking this from being merged since I want to ensure the e2e tests won't be broken by this change before this is merged.

File: datalake-dr-connector/src/main/java/com/sequenceiq/cloudbreak/datalakedr/DatalakeDrClient.java
Comment (RokolyA94) [inline datalake-dr-connector/src/main/java/com/sequenceiq/cloudbreak/datalakedr/DatalakeDrClient.java:319]: the `else` is not required.

```java
   private ManagedChannelWrapper makeWrapper(boolean mockCloud) {
        if (mockCloud) {
            return new ManagedChannelWrapper(
                    ManagedChannelBuilder.forAddress(datalakeDrConfig.getMockHost(), datalakeDrConfig.getMockPort())
                            .usePlaintext()
                            .maxInboundMessageSize(DEFAULT_MAX_MESSAGE_SIZE)
                            .build());
        }
        return new ManagedChannelWrapper(
        ManagedChannelBuilder.forAddress(datalakeDrConfig.getHost(), datalakeDrConfig.getPort())
                .usePlaintext()
                .maxInboundMessageSize(DEFAULT_MAX_MESSAGE_SIZE)
                .build());
  }

File: datalake-dr-connector/src/main/java/com/sequenceiq/cloudbreak/datalakedr/config/DatalakeDrConfig.java
Comment (RokolyA94) [inline datalake-dr-connector/src/main/java/com/sequenceiq/cloudbreak/datalakedr/config/DatalakeDrConfig.java:77]: We can overload the `isConfigured()` method to provide an "old" solution that can call the new one. in that case, you do not need to change all possible locations of the usage of this method.

e.q:
```java 
public boolean isConfigured(){
   return isConfigured(true);
}
```

**Review by RokolyA94 — APPROVED** [review]

File: datalake-dr-connector/src/main/java/com/sequenceiq/cloudbreak/datalakedr/DatalakeDrClient.java
Comment (sxxgrc) [reply datalake-dr-connector/src/main/java/com/sequenceiq/cloudbreak/datalakedr/DatalakeDrClient.java:319]: True thank you I'll change this.

File: datalake-dr-connector/src/main/java/com/sequenceiq/cloudbreak/datalakedr/config/DatalakeDrConfig.java
Comment (sxxgrc) [reply datalake-dr-connector/src/main/java/com/sequenceiq/cloudbreak/datalakedr/config/DatalakeDrConfig.java:77]: Yeah this may be cleaner thank you.

**Comment by sxxgrc** [comment]
I believe I am okay with moving forward and merging this in with the understanding that the nightly e2e SDX resize test may fail due to the environment not being configured properly to run the Datalake DR stuff.

I have created a branch which will revert the problematic checks which would lead to any issues here: https://github.com/hortonworks/cloudbreak/tree/revert_CB-17317_e2e_test 
I have split up the change there into 2 new commits: one reverting the e2e test and one reverting the mock test as I am unsure of which may break the nightly tests.

File: datalake-dr-connector/src/main/java/com/sequenceiq/cloudbreak/datalakedr/config/DatalakeDrConfig.java
Comment (sxxgrc) [reply datalake-dr-connector/src/main/java/com/sequenceiq/cloudbreak/datalakedr/config/DatalakeDrConfig.java:81]: Actually looking at this again, this would only be called in 3 places with 2 of those being tests so I don't think it's necessary.

Comment (RokolyA94) [reply datalake-dr-connector/src/main/java/com/sequenceiq/cloudbreak/datalakedr/config/DatalakeDrConfig.java:81]: sorry, in that case just ignore my comments!

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxBackupRestoreService.java
Comment (kkalvagadda1) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxBackupRestoreService.java:183]: Can you elaborate on this change? Looks like a non-related change.

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/recovery/SdxUpgradeRecoveryService.java
Comment (kkalvagadda1) [inline datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/recovery/SdxUpgradeRecoveryService.java:70]: Why is the change made?

**Comment by horadla23** [comment]
to be honest, I do not get this check. you have introduced different fields for mock, but you should be able to configure different values for host,port in case of mock, so I do not understand why we need a different value for that

**Comment by sxxgrc** [comment]
> to be honest, I do not get this change. you have introduced different fields for mock, but you should be able to configure different values for host,port in case of mock (in integration-test module or in jenkins jobs where mock tests are initialized and executed), so I do not understand why we need a different value for that

Sorry that it wasn't clear.

My main goal with the changes was to ensure that when a mock cloud provider runs backup/restore operations, it correctly communicates with the mock Datalake DR service which currently runs as part of the mock thunderhead service in cloudbreak. 
So my changes are meant to do the following:

- Validate backup/restore operations within the mock resize tests
- Dynamically switch between the real Datalake DR endpoint and the mock one according to whether the cloud provider is mocked or not

I'm assuming you're asking about this last point right? As we could just switch out the original DR endpoint to the mock one when needed. I guess my original reason for not doing this was thinking that we should be able to use both the real DR endpoint and the mock one within the same environment, but from what I've heard lately there isn't really an argument for this being necessary. I don't believe both the mock tests and e2e tests are ever run within the same testing environment right?

If there is no point where this dynamic switching would be useful then I will make the modifications to simplify this change and expect a manual override of the original endpoint.

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/recovery/SdxUpgradeRecoveryService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/recovery/SdxUpgradeRecoveryService.java:70]: This was really just to remove the unnecessary uses of the DrClient and just use the backuprestoreservice but this isn't a necessary change.

**Comment by sxxgrc** [comment]
I have simplified the code change a lot per my previous comment.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxBackupRestoreService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxBackupRestoreService.java]: See my comment on your other comment.

File: integration-test/integcb/Profile_template
Comment (kkalvagadda1) [inline integration-test/integcb/Profile_template:32]: Is this file used only for mock tests?

**Review by kkalvagadda1 — COMMENTED** [review]
I see below changes to the config
1. Enable Dr bu default in datalake service
2. Exposing 8981 for thunder mock service
3. Enable DATALAKE_DR_ENABLED for tests
4. configuring thunderhead-mock:8981 as dr endpoint

Does that mean the only config change that is relevant for integration tests is item-1?
Can you confirm?

File: integration-test/integcb/Profile_template
Comment (sxxgrc) [reply integration-test/integcb/Profile_template:32]: Yes this is used only for the tests run in the PR which all use the mock cloud provider.

**Comment by sxxgrc** [comment]
> I see below changes to the config
> 
>     1. Enable Dr bu default in datalake service
> 
>     2. Exposing 8981 for thunder mock service
> 
>     3. Enable DATALAKE_DR_ENABLED for tests
> 
>     4. configuring thunderhead-mock:8981 as dr endpoint
> 
> 
> Does that mean the only config change that is relevant for integration tests is item-1? Can you confirm?

Yes you are correct, the other three are just for the mock tests.

**Review by kkalvagadda1 — APPROVED** [review]

**Review by horadla23 — APPROVED** [review]

**Pull request merged**
